Homework 2

a-s01 opened the pull request.

a-s01 committed:
HW_2: implement beans and config classes showing different lifecycle management approaches

Resolves: task items 1-9

a-s01 committed:
HW_2: add BeanPostProcessor and BeanFactoryPostProcessor implementations

Validation is based on annotations. @ValidationRequired marks bean or annotation
for further validation, @IsNotNull and @IsInRange for fields show the condition
to match, and according Validator implementations do the work. Check validation
package

Resolves: task items 10-12

a-s01 committed:
HW_2: update: change exception throwing to console message log due to new task input

Also, now beanA doesn't pass validation for feature demonstration

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BeanA.java:
```diff
@@ -0,0 +1,42 @@
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Value;
```
unused import

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BeanF.java:
```diff
@@ -0,0 +1,35 @@
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.factory.annotation.Autowired;
```
unused import

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/ValidationRequiredAnnotationBeanPostProcessor.java:
```diff
@@ -0,0 +1,67 @@
+import com.epam.spring.homework2.validation.ValidatorFactory;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanInitializationException;
```
unused import

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BeanB.java:
```diff
@@ -0,0 +1,42 @@
+public class BeanB {
+
+    @IsNotNull
+    @Value("${beanB.name}")
```
This is not a good approach. Please rework:
- make class immutable
- inject values from properties file in a method inside configuration class

Apply to all relevant classes

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/Main.java:
```diff
@@ -0,0 +1,19 @@
+        for (String beanDefinitionName : context.getBeanDefinitionNames()) {
+            System.out.println("- " + context.getBean(beanDefinitionName));
+        }
+
```
Missing item No 12

```java
System.out.println("\nAll beans definitions present in application context: ");
for (String beanDefinitionName : context.getBeanDefinitionNames()) {
    System.out.println("- " + context.getBeanDefinition(beanDefinitionName));
}
```

KostyantynPanchenko reviewed (changes requested):
Please make fixes according to the comments.

Also configure `gitignore` to exclude `*.iml` and binaries. Examples are available in internet.

a-s01 committed:
update gitignore

a-s01 committed:
HW_2: fix: make BeanB-D immutable and configured accordingly, add missing task item 12, delete unused imports

a-s01 commented:
Thanks for the review. 
I have fixed my issues. Please, check.

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BeanC.java:
```diff
@@ -0,0 +1,38 @@
+    public void destroyC() {
+        System.out.println("beanC: destroyC()");
+    }
+
```
extra line between the methods

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BeanD.java:
```diff
@@ -0,0 +1,38 @@
+    public void destroyD() {
+        System.out.println("beanD: destroyD()");
+    }
+
```
extra line between the methods

KostyantynPanchenko commented on .gitignore:
```diff
@@ -1,2 +1,23 @@
+*.class
+
+# Log file
+*.log
```
Do you also see that red sign over here? :) It requires you to add a new line at the end of the file :)

a-s01 replied:
This just sucks. Sorry for it. I enabled the conforming setting in Intelijii, it should help me get rid of these forever. At least, I hope so :)

KostyantynPanchenko commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+beanC.name = C
+beanC.value = 2
+beanD.name = D
+beanD.value = 3
```
And here...

KostyantynPanchenko reviewed (changes requested):
Please make changes according to the comments and have a great weekend

a-s01 committed:
HW_2: fix: cosmetic changes

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/config/Config.java:
```diff
@@ -0,0 +1,45 @@
+    @Value("${beanD.value}")
+    private int beanDValue;
+
+    @Bean(initMethod = "initD", destroyMethod = "destroyD")
```
Let's make the world beautiful with
```java
@Bean(initMethod = "initD", destroyMethod = "destroyD")
    public BeanD beanD(@Value("${beanD.name}") final String name, @Value("${beanD.value}") final int value) {
        return new BeanD(name, value);
    }
```
And get rid of those instance variables.

If we go even further - we can introduce a base class with getters, `toString()` and both `init` and `destroy` methods. Then we can make `BeanB`, `BeanC` and `BeanD` extend our base class ;)

a-s01 commented:
@KostyantynPanchenko would you mind checking this also? My bad, I didn't comment on my done work.

KostyantynPanchenko commented:
I am talking about this comment:
![Screenshot 2022-02-07 225327](https://user-images.githubusercontent.com/17402170/152870249-50d4d23c-1899-47db-9003-1560df5ec81c.jpeg)

a-s01 committed:
HW_2: make the world beautiful :) change Config.java

a-s01 committed:
HW_2: add BaseBean super class for BeanB-D

I wanted to make the child classes as small as possible, so changes
in the BeanPostProcessor was required. Now it chechs not only bean's
fields but also its super class fields.

Was cool excersise :)

a-s01 commented:
It's strange I didn't see the requested changes. Could it be because it was pending?
Anyway, thanks for the review!
![Screenshot 2022-02-08 at 09 40 57](https://user-images.githubusercontent.com/1070584/152940405-188069be-0b51-4e78-8fbf-d1621d2471b6.png)

KostyantynPanchenko commented on src/main/java/com/epam/spring/homework2/bean/BaseBean.java:
```diff
@@ -0,0 +1,31 @@
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+
+public abstract class BaseBean {
```
Does not contain abstract methods, thus should not be abstract. Also most likely can be package-private

KostyantynPanchenko reviewed (commented):
See my comment in BaseBean

a-s01 committed:
HW_2: fix: make BaseBean not abstract & package private
